### PR TITLE
BFD-798: Amazon Linux 2 AMI Migration

### PR DESCRIPTION
**Pull request by njdister**
**JIRA Ticket:**
[BFD-798](https://jira.cms.gov/browse/BFD-798)

**User Story or Bug Summary:**
As a BFD team engineer, I want to transition from Redhat to AL2 in order to move the BFD away from the Gold image, so that there is time saved trouble shooting the image, performance improvements with the new image, and increased security with the new software.

### What Does This PR Do?
This PR makes the following changes to migrate AMIs to AL2:
- Update platinum playbook to remove RH7 specific steps and consolidate dependency installation
- Update bfd-server and bfd-pipeline roles to remove RH7 specific steps
- Update packer files to namespace AL2 AMI builds and set equal disk space to RH7 AMIs
- Update deploy Jenkins stages to use namespaced AL2 AMIs and AL2 gold image convention from ITOPS
- Update terraform to support new AL2 AMI disk device (/dev/xvda) and remove RH7 specific disk extension

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:
- Is this a good approach (wholesale replace RH7) or should we support AL2 deployments as a switch while fully supporting RH7 just in case?  We did this for BB2 but ended up not needing it.  Open to discussion.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [x] Yes
      * [ ] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [x] Yes
      * [ ] No

Important note, we have an existing AL2 SIA here: https://confluence.cms.gov/display/BFD/2021-05-14+Amazon+Linux+2

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: ops/ansible/playbooks-ccs/build_bfd-platinum.yml [hunk 1]
```diff
@@ -12,15 +12,13 @@
...
-    - import_role:
-        name: platinum_post
+    - name: Install dependencies
+      yum:
+        name:
+          - java-1.8.0-openjdk-devel
+          - git
+          - ansible
+          - amazon-cloudwatch-agent
+          - amazon-efs-utils
```
Comment (njdister) [inline ops/ansible/playbooks-ccs/build_bfd-platinum.yml:24]: Just cutting out the middle man here due to simplicity, if this is not desired I can create a new role for AL2 dependencies.

File: ops/ansible/roles/bfd-pipeline/tasks/eft_efs.yml [hunk 2]
```diff
@@ -9,18 +9,6 @@
...
-    name: ../../playbooks-ccs/roles/amazon-efs-utils
-  tags:
-    - pre-ami
-
-- name: Build fips compatible stunnel from source
-  import_role:
-    name: ../../playbooks-ccs/roles/stunnel
-  tags:
-    - pre-ami
-
```
Comment (njdister) [inline ops/ansible/roles/bfd-pipeline/tasks/eft_efs.yml:23]: No longer required in AL2

File: ops/ansible/roles/bfd-pipeline/tasks/etl.yml [hunk 3]
```diff
@@ -1,16 +1,5 @@
...
-  yum:
-    name: "{{ item }}"
-    state: present
-  with_items:
-    # Needed to run the ETL service.
-    - java-1.8.0-openjdk
-  become: true
-  tags: 
-    - pre-ami
-
```
Comment (njdister) [inline ops/ansible/roles/bfd-pipeline/tasks/etl.yml:13]: Java is already installed during the platinum build, not required here.

File: ops/ansible/roles/bfd-server/tasks/install.yml [hunk 4]
```diff
@@ -1,14 +1,5 @@
...
 
-- name: Install Pre-requisites
-  yum:
-    name: "{{ item }}"
-    state: present
-  with_items:
-    # Needed to supply `keytool` and also to run the app server.
-    - java-1.8.0-openjdk-devel
-  become: true
-
```
Comment (njdister) [inline ops/ansible/roles/bfd-server/tasks/install.yml:11]: Java is already installed during the platinum build, not required here.

File: ops/ansible/roles/bfd-server/tasks/install.yml [hunk 5]
```diff
@@ -114,14 +105,3 @@
...
-- name: Add selinux file context to allow log rotation
-  sefcontext:
-    target: "{{ data_server_dir }}(/.*json|/.*log)?"
-    setype: var_log_t
-    state: present
-  become: true
-
-- name: Apply selinux file context
-  command: "restorecon -rv {{ data_server_dir }}"
-  become: true
```
Comment (njdister) [inline ops/ansible/roles/bfd-server/tasks/install.yml:127]: No longer required in AL2.

File: ops/ansible/roles/cloudwatch-agent-instrumented/tasks/main.yml [hunk 6]
```diff
@@ -12,15 +12,6 @@
...
 
-- name: cwagent | install agent from RPM
-  yum:
-    name: https://s3.amazonaws.com/amazoncloudwatch-agent/redhat/amd64/latest/amazon-cloudwatch-agent.rpm
-    state: present
-    disable_gpg_check: true
-  become: true
-  tags:
-    - pre-ami
-
```
Comment (njdister) [inline ops/ansible/roles/cloudwatch-agent-instrumented/tasks/main.yml:23]: Installed during the platinum build so no longer required.

File: ops/deploy-ccs.groovy [hunk 7]
```diff
@@ -51,21 +51,21 @@ def findAmis() {
...
-			'Name=name,Values=bfd-etl-??????????????' \
+			'Name=name,Values=bfd-amzn2-etl-??????????????' \
 			'Name=state,Values=available' --region us-east-1 --output json | \
 			jq -r '.Images | sort_by(.CreationDate) | last(.[]).ImageId'"
     ).trim(),
     bfdServerAmiId: sh(
       returnStdout: true,
       script: "aws ec2 describe-images --owners self --filters \
-			'Name=name,Values=bfd-fhir-??????????????' \
+			'Name=name,Values=bfd-amzn2-fhir-??????????????' \
```
Comment (njdister) [inline ops/deploy-ccs.groovy:68]: Namespace AL2 AMIs for auditing and easy rollback if required.

File: ops/deploy-ccs.groovy [hunk 8]
```diff
@@ -90,7 +90,7 @@ def buildPlatinumAmi(AmiIds amiIds) {
 		def goldAmi = sh(
 			returnStdout: true,
 			script: "aws ec2 describe-images --filters \
-			'Name=name,Values=\"EAST-RH 7-? Gold Image V.1.?? (HVM) ??-??-??\"' \
+			'Name=name,Values=\"amzn2legacy*\"' \
```
Comment (njdister) [inline ops/deploy-ccs.groovy:93]: Swap RHEL7 gold image finder for AL2 gold image finder.

File: ops/packer/build_bfd-all.json [hunk 9]
```diff
@@ -47,10 +47,10 @@
           "tag:Name": "bfd-mgmt-vpn-private"
         }
       },
-      "ami_name": "bfd-fhir-{{isotime \"20060102030405\"}}",
+      "ami_name": "bfd-amzn2-fhir-{{isotime \"20060102030405\"}}",
       "ssh_pty": true,
       "tags": {
-        "Name": "bfd-fhir-{{isotime \"20060102030405\"}}",
+        "Name": "bfd-amzn2-fhir-{{isotime \"20060102030405\"}}",
```
Comment (njdister) [inline ops/packer/build_bfd-all.json:53]: Namespace AL2 AMIs for auditing and easy rollback if required.

File: ops/packer/build_bfd-platinum.json [hunk 10]
```diff
@@ -18,10 +18,18 @@
...
+      "ami_name": "bfd-amzn2-platinum-{{isotime \"20060102030405\"}}",
       "ssh_pty": true,
+      "launch_block_device_mappings": [
+        {
+          "device_name": "/dev/xvda",
+          "volume_type": "gp2",
+          "volume_size": 30,
+          "delete_on_termination": true
+        }
+      ],
```
Comment (njdister) [inline ops/packer/build_bfd-platinum.json:30]: Ensure our platinum image has the same root volume size as RHEL 7 gold image did.

File: ops/terraform/modules/resources/asg/main.tf [hunk 11]
```diff
@@ -94,7 +94,7 @@ resource "aws_launch_template" "main" {
   }
 
   block_device_mappings {
-    device_name = "/dev/sda1"
+    device_name = "/dev/xvda"
```
Comment (njdister) [inline ops/terraform/modules/resources/asg/main.tf:97]: Update root volume nomenclature to work with AL2.  Strangely this was not required for the ETL build (EC2 builder seems to auto find correct device name).

File: ops/terraform/modules/resources/templates/fhir_server.tpl [hunk 12]
```diff
@@ -8,14 +8,6 @@ exec > >(
...
 )
 
-# Extend gold image defined root partition with all available free space
-# The extend gold image has been commented out in favor of rotating the logs
-# Doing this should improve deploy time, BLUEBUTTON-1582
-growpart /dev/nvme0n1 2
-pvresize /dev/nvme0n1p2
-lvextend -l +100%FREE /dev/VolGroup00/rootVol
-xfs_growfs /
-
```
Comment (njdister) [inline ops/terraform/modules/resources/templates/fhir_server.tpl:18]: No longer required in AL2, root volume is expanded automagically upon defining extra space.

File: ops/terraform/modules/resources/templates/pipeline_server.tpl [hunk 13]
```diff
@@ -8,12 +8,6 @@ exec > >(
 	done
 )
 
-# Extend gold image defined root partition with all available free space
-sudo growpart /dev/nvme0n1 2
-sudo pvresize /dev/nvme0n1p2
-sudo lvextend -l +100%FREE /dev/VolGroup00/rootVol
-sudo xfs_growfs /
-
```
Comment (njdister) [inline ops/terraform/modules/resources/templates/pipeline_server.tpl:16]: No longer needed in AL2, root volume is expanded automagically upon defining extra space.

**Review by jzulim — APPROVED** [review]
Great work. LGTM. Thank you for this.

File: ops/deploy-ccs.groovy (hunk 7 above)
Comment (keithdadkins) [reply ops/deploy-ccs.groovy:68]: I verified there are recently updated gold AL2 images:
```
"2021-08-31T22:23:40.000Z"
"2021-09-01T12:53:08.000Z"
"2021-09-01T13:59:46.000Z"
"2021-09-07T16:38:07.000Z"
```

**Review by keithdadkins — APPROVED** [review]
LGTM. This work will dramatically help speed up build and deployment times, improve security, reduces dependencies, and may actually perform a bit better in prod. Good stuff!

**Review by StewGoin — APPROVED** [review]
This looks in order to me, go forth and speed up deploys!

File: ops/ansible/roles/bfd-server/tasks/install.yml (hunk 4 above)
Comment (karlmdavis) [reply ops/ansible/roles/bfd-server/tasks/install.yml:11]: FYI: Undoing this as part of the capsule changes. Two reasons:

1. We have tests that also use the role.
2. It's not listed as a pre-req for the role, so the role should provide it.

File: ops/ansible/roles/bfd-pipeline/tasks/etl.yml (hunk 3 above)
Comment (karlmdavis) [reply ops/ansible/roles/bfd-pipeline/tasks/etl.yml:13]: FYI: Undoing this as part of the capsule changes. Two reasons:

1. We have tests that also use the role.
2. It's not listed as a pre-req for the role, so the role should provide it.

**Pull request merged**